### ISSUE TITLE
Proposal: Predicate inferer

### ISSUE THREAD

**Issue by emmbm** [sindresorhus/type-fest#973]
### Type description + examples

With typescript now supporting predicate inference for guard functions without the need for explicitly specifying a `is` clause, it could be interesting to provide a simple helper that allows to get the inferred predicate for any guard function.

Ex.:
```ts
function isValid(value: string | number | null | undefined | false) {
  return value != null && value !== false;
}

type Valid = Predicate<typeof isValid> // string | number
```

### Type source

```ts
export type Predicate<T> = T extends (x: any) => x is infer U ? U : never;
```

### Search existing types and issues first

- [X] I tried my best to look for it

**Comment by sindresorhus** [comment]
Sounds useful, but the name is slightly confusing. The type extracts the inferred guard type, but a name like `Predicate` makes it sound like it's for making predicates. Maybe something like `GuardType` or `ExtractGuard`?

**Comment by emmbm** [comment]
I agree `Predicate` is probably not the best naming, but I don't think derivatives of `Guard` in the present tense sound right either since the guard is technically the check, not it's resulting assertion. Maybe `GuardedType`?

**Comment by fregante** [comment]
This type would be similar to `ReturnType` so maybe it should follow its naming.

**Comment by Emiyaaaaa** [comment]
Looks interesting, are there any use cases?

**Comment by emmbm** [comment]
@Emiyaaaaa rather than having to rely on manually maintaining consistency across explicit predicates and implementations, it allows to derive types from guard function implementations, the latter being less error-prone.

For example:

```ts
// Before

type Valid = string | number | null // Let's say a dev decided to add null to this union.

/**
 * When declared explicitly, the predicate can diverge from the actual implementation if not cautious.
 */
function isValid(value: unknown): value is Valid {
  return value != null && value !== false;
}
```
```ts
// After

function isValid(value: unknown) {
  return value != null && value !== false;
}

type Valid = Predicate<typeof isValid> // This will always be up to date with the guard implementation.
```

**Comment by Emiyaaaaa** [comment]
Sorry I actually want real usecases, and I have a doubt that this type depends on the real result of the execution of the function, this sounds a little strange for typescript.

**Comment by emmbm** [comment]
> and I have a doubt that this type depends on the real result of the execution of the function, this sounds a little strange for typescript.

I'm not sure how to interpret this, it honestly reads either like a misinterpretation conflating runtime and compile time analysis, or an overlook of predicate inference being a major improvement brought by typescript 5.5 (https://github.com/microsoft/TypeScript/pull/57465).

Either way, if you don't think this would be a valid addition, feel free to close the issue.
